nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/services/impl/GlassfishNetworkListener.java

```diff
@@ -437,24 +447,35 @@ protected boolean onHttpHeaderParsed(final HttpHeader httpHeader,
             if (serverVersion != null && !serverVersion.isEmpty()) {
                 response.addHeader(Header.Server, serverVersion);
             }
-
-
-            
+     
             // Set response "X-Powered-By" header
             if (xPoweredBy != null) {
                 response.addHeader(Header.XPoweredBy, xPoweredBy);
             }
             
+            // Set response "X-Frame-Options" header
+            if (xFrameOptions != null) {
+                response.addHeader(xFrameOptionsHeader, xFrameOptions);
+            }
+            
             return result;
         }
         
-        // override to chec whether x-Powered By has been added by something else
+        // Override to check whether x-Powered By, Server header and x-Frame-Option has been added by something else
         @Override
         protected void onInitialLineEncoded(HttpHeader httpHeader, FilterChainContext ctx) {
           
             if (xPoweredBy == null && httpHeader.containsHeader(Header.XPoweredBy)) {
                 httpHeader.getHeaders().removeHeader(Header.XPoweredBy);
             }
+            
+            if (serverVersion == null && httpHeader.containsHeader(Header.Server)){
+                httpHeader.getHeaders().removeHeader(Header.Server);
+            }
+            
+            if (xFrameOptions == null && httpHeader.containsHeader(xFrameOptionsHeader)){
```

Comment:
I don't think you should remove an XFrameOptions Header if it is already present. XPoweredBy and Server are removed when configured to be removed because they leak security information. However XFrame-Options could be set by the application developer for legitimate reasons so should be honoured if present.